File: src/net/liplum/api/cyber/IDataSender.java
package net.liplum.api.cyber;

import arc.struct.ObjectSet;
import mindustry.type.Item;
import net.liplum.SendDataPack;
import net.liplum.utils.ArcU;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface IDataSender extends IDataBuilding {
    /**
     * sends items
     *
     * @param receiver the target who receives the item(s)
     * @param item     which kind of item will be sent soon
     * @param amount   how many item(s) will be sent
     * @return the rest of item(s)
     */
    default int sendData(@NotNull IDataReceiver receiver, @NotNull Item item, int amount) {
        int maxAccepted = receiver.acceptedAmount(this, item);
        if (maxAccepted == -1) {
            receiver.receiveData(this, item, amount);
            return 0;
        }
        if (maxAccepted >= amount) {
            receiver.receiveData(this, item, amount);
            return 0;
        } else {
            int rest = amount - maxAccepted;
            receiver.receiveData(this, item, maxAccepted);
            return rest;
        }
    }

    @SendDataPack
    void connectSync(@NotNull IDataReceiver receiver);

    @SendDataPack
    void disconnectSync(@NotNull IDataReceiver receiver);

    @SendDataPack
    default void connectSync(int receiver) {
        IDataReceiver dr = CyberH.dr(receiver);
        if (dr != null) {
            connectSync(dr);
        }
    }

    @SendDataPack
    default void disconnectSync(int receiver) {
        IDataReceiver dr = CyberH.dr(receiver);
        if (dr != null) {
            disconnectSync(dr);
        }
    }

    @Nullable
    Integer connectedReceiver();

    default boolean canMultipleConnect() {
        return maxReceiverConnection() != 1;
    }

    default boolean isConnectedWith(@NotNull IDataReceiver receiver) {
        if (canMultipleConnect()) {
            return connectedReceivers().contains(receiver.getBuilding().pos());
        } else {
            Integer connected = connectedReceiver();
            if (connected == null) {
                return false;
            } else {
                return connected == receiver.getBuilding().pos();
            }
        }
    }

    /**
     * Gets the maximum limit of connection.<br/>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    default int maxReceiverConnection() {
        return 1;
    }

    default boolean canHaveMoreReceiverConnection() {
        int max = maxReceiverConnection();
        if (max == -1) {
            return true;
        }
        return connectedReceivers().size < max;
    }

    default int getReceiverConnectionNumber() {
        if (canMultipleConnect()) {
            return connectedReceivers().size;
        } else {
            return connectedReceiver() == null ? 0 : 1;
        }
    }

    @NotNull
    default ObjectSet<Integer> connectedReceivers() {
        return ArcU.emptySet();
    }
}

File: main/src/net/liplum/api/cyber/IDataSender.kt
package net.liplum.api.cyber

import arc.graphics.Color
import arc.struct.ObjectSet
import mindustry.type.Item
import net.liplum.R
import net.liplum.api.ICyberEntity
import net.liplum.mdt.ClientOnly
import net.liplum.mdt.SendDataPack

interface IDataSender : ICyberEntity {
    /**
     * sends items
     *
     * @param receiver the target who receives the item(s)
     * @param item     which kind of item will be sent soon
     * @param amount   how many item(s) will be sent
     * @return the rest of item(s)
     */
    fun sendDataTo(receiver: IDataReceiver, item: Item, amount: Int): Int {
        val maxAccepted = receiver.getAcceptedAmount(this, item)
        if (maxAccepted == -1) {
            receiver.receiveDataFrom(this, item, amount)
            return 0
        }
        return if (maxAccepted >= amount) {
            receiver.receiveDataFrom(this, item, amount)
            0
        } else {
            val rest = amount - maxAccepted
            receiver.receiveDataFrom(this, item, maxAccepted)
            rest
        }
    }
    @SendDataPack
    fun connectToSync(receiver: IDataReceiver)
    @SendDataPack
    fun disconnectFromSync(receiver: IDataReceiver)
    fun isConnectedTo(receiver: IDataReceiver): Boolean {
        return receiver.building.pos() in connectedReceivers
    }
    /**
     * Gets the maximum limit of connection.<br></br>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    val maxReceiverConnection: Int
        get() = 1
    val canHaveMoreReceiverConnection: Boolean
        get() = maxReceiverConnection == -1 || receiverConnectionNumber < maxReceiverConnection
    val receiverConnectionNumber: Int
        get() = connectedReceivers.size
    @ClientOnly
    val senderColor: Color
        get() = R.C.Sender
    val maxRange: Float
        get() = -1f
    val connectedReceivers: ObjectSet<Int>

    companion object {
        @SendDataPack
        fun IDataSender.connectToSync(receiver: Int) {
            receiver.dr()?.let { connectToSync(it) }
        }
        @SendDataPack
        fun IDataSender.disconnectFromSync(receiver: Int) {
            receiver.dr()?.let { disconnectFromSync(it) }
        }
    }
}